lib/std/fmt/parse_float/convert_eisel_lemire.zig

```diff
@@ -0,0 +1,843 @@
+        const inside_safe_exponent = q >= -27 and q <= 55;
+        if (!inside_safe_exponent) {
+            return null;
+        }
+    }
+
+    const upper_bit = @intCast(i32, r.hi >> 63);
+    var mantissa = math.shr(u64, r.hi, upper_bit + 64 - @intCast(i32, float_info.mantissa_explicit_bits) - 3);
+    var power2 = power(@intCast(i32, q)) + upper_bit - @intCast(i32, lz) - float_info.minimum_exponent;
+    if (power2 <= 0) {
+        if (-power2 + 1 >= 64) {
+            // Have more than 64 bits below the minimum exponent, must be 0.
+            return BiasedFp(f64).zero();
+        }
+        // Have a subnormal value.
+        mantissa = math.shr(u64, mantissa, -power2 + 1);
+        mantissa += mantissa & 1;
+        mantissa >>= 1;
+        power2 = @boolToInt(mantissa >= (1 << float_info.mantissa_explicit_bits));
+        return BiasedFp(f64){ .f = mantissa, .e = power2 };
+    }
+
+    // Need to handle rounding ties. Normally, we need to round up,
+    // but if we fall right in between and and we have an even basis, we
+    // need to round down.
+    //
+    // This will only occur if:
+    //  1. The lower 64 bits of the 128-bit representation is 0.
+    //      IE, 5^q fits in single 64-bit word.
+    //  2. The least-significant bit prior to truncated mantissa is odd.
+    //  3. All the bits truncated when shifting to mantissa bits + 1 are 0.
+    //
+    // Or, we may fall between two floats: we are exactly halfway.
+    if (r.lo <= 1 and
+        q >= float_info.min_exponent_round_to_even and
+        q <= float_info.max_exponent_round_to_even and
+        mantissa & 3 == 1 and
+        math.shl(u64, mantissa, (upper_bit + 64 - @intCast(i32, float_info.mantissa_explicit_bits) - 3)) == r.hi)
+    {
+        // Zero the lowest bit, so we don't round up.
+        mantissa &= ~@as(u64, 1);
+    }
+
+    // Round-to-even, then shift the significant digits into place.
+    mantissa += mantissa & 1;
+    mantissa >>= 1;
+    if (mantissa >= 2 << float_info.mantissa_explicit_bits) {
+        // Rounding up overflowed, so the carry bit is set. Set the
+        // mantissa to 1 (only the implicit, hidden bit is set) and
+        // increase the exponent.
+        mantissa = 1 << float_info.mantissa_explicit_bits;
+        power2 += 1;
+    }
+
+    // Zero out the hidden bit
+    mantissa &= ~(@as(u64, 1) << float_info.mantissa_explicit_bits);
+    if (power2 >= float_info.infinite_power) {
+        // Exponent is above largest normal value, must be infinite
+        return BiasedFp(f64).inf(T);
+    }
+
+    return BiasedFp(f64){ .f = mantissa, .e = power2 };
+}
+
+/// Calculate a base 2 exponent from a decimal exponent.
+/// This uses a pre-computed integer approximation for
+/// log2(10), where 217706 / 2^16 is accurate for the
+/// entire range of non-finite decimal exponents.
+fn power(q: i32) i32 {
+    return ((q *% (152170 + 65536)) >> 16) + 63;
+}
+
+const U128 = struct {
```

Comment:
We don't need this type and should use `u128` directly. I will likely follow up with this change after measuring performance.